tools/pubsys-setup/src/main.rs

```diff
@@ -0,0 +1,384 @@
+        });
+    }
+}
+
+/// Main entry point for tuftool setup.
+fn run() -> Result<()> {
+    // Parse and store the args passed to the program
+    let args = Args::from_args();
+
+    // TerminalMode::Mixed will send errors to stderr and anything less to stdout.
+    TermLogger::init(args.log_level, LogConfig::default(), TerminalMode::Mixed)
+        .context(error::Logger)?;
+
+    // Main branching logic for deciding whether to create role/key, use what we have, or error.
+    match find_root_role_and_key(&args)? {
+        (Some(_root_role_path), Some(_key_url)) => Ok(()),
+        (Some(_root_role_path), None) => {
+            ensure!(
+                args.allow_missing_key,
+                error::MissingKey { repo: args.repo }
+            );
+            Ok(())
+        }
+        // User is missing something, so we generate at least a root.json and maybe a key.
+        (None, maybe_key_url) => {
+            if maybe_key_url.is_some() {
+                info!("Didn't find root role in Infra.toml, generating...");
+            } else {
+                info!("Didn't find root role or signing key in Infra.toml, generating...");
+            }
+
+            let role_dir = args.root_role_path.parent().context(error::Path {
+                path: &args.root_role_path,
+                thing: "root role",
+            })?;
+            fs::create_dir_all(role_dir).context(error::Mkdir { path: role_dir })?;
+
+            let temp_path = NamedTempFile::new_in(&role_dir).context(error::TempFileCreate {
+                purpose: "root role",
+            })?;
+            let temp_path_display = temp_path.path().display();
+
+            // Make tuftool calls to create an initial root.json with basic parameters.
+            tuftool!("root init '{}'", temp_path_display);
+
+            tuftool!("root expire '{}' 'in 52 weeks'", temp_path_display);
+
+            tuftool!("root set-threshold '{}' root 1", temp_path_display);
+            tuftool!("root set-threshold '{}' snapshot 1", temp_path_display);
+            tuftool!("root set-threshold '{}' targets 1", temp_path_display);
+            tuftool!("root set-threshold '{}' timestamp 1", temp_path_display);
+
+            let key_url = if let Some(key_url) = maybe_key_url {
+                // If the user has a key, add it to each role.
+                tuftool!("root add-key '{}' '{}' --role root --role snapshot --role targets --role timestamp",
+                         temp_path_display, key_url);
+                key_url
+            } else {
+                // If the user has no key, build one and add it to each role.
+                let key_dir = args.default_key_path.parent().context(error::Path {
+                    path: &args.default_key_path,
+                    thing: "key",
+                })?;
+                fs::create_dir_all(key_dir).context(error::Mkdir { path: key_dir })?;
```

Comment:
nit: if we did this `mkdir` somewhere else, it would make the two branches a bit more symmetrical.

Also, I know it's wasteful, but there'd be a lot less branching if we generated the key in `find_root_role_and_key`, possibly by calling `tuftool root gen-rsa-key ...` with a temp file and throwing away the root.json we don't need.

We could wrap the initial few `tuftool` commands in a function.

Reply:
Our push moved the mkdirs, which does clarify the code, but we didn't think it made as much sense to generate a key inside the find function.  It seems unexpected for that to happen given the purpose of that function.  The conditional in the `run` function that handles key generation is relatively small, especially now that the mkdir is moved.  It seems to be pretty clear now.